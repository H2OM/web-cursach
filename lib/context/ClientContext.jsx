'use client';
import { createContext, useEffect, useState } from "react";

const ClientContext = createContext();

export const ClientProvider = ({children}) => {
   const [notification, setNotification] = useState({message: false, status: true, closing: false });
   useEffect(()=>{
        if(notification.message && !notification.closing) {
            setNotification((current)=>({...current, closing: true}));
            setTimeout(()=>setNotification({message: false, status: true, closing: false}), 4000);
        }
    }, [notification]);

    return (
        <ClientContext.Provider
            value={{
                setNotification,
                notification
            }}
        >
            {children}
        </ClientContext.Provider>
    )
}

export default ClientContext;