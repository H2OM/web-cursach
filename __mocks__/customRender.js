import ClientContext from '@/lib/context/ClientContext';
import { render } from '@testing-library/react';

export default customRender = (ui, {providerProps}) => {
    return render(
      <ClientContext.Provider value={{...providerProps}}>{ui}</ClientContext.Provider>,
    )
}