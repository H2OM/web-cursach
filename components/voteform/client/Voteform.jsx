'use client';
import Loader from '@/lib/basecomponents/loader/loader';
import '../voteform.scss';
import ClientContext from "@/lib/context/ClientContext";
import { useContext, useEffect, useState } from "react";


export default function Voteform () {
    const {voteMenu, setVoteMenu, setNotification} = useContext(ClientContext);
    const [status, setStatus] = useState(false);
    const [select, setSelect] = useState(false);

    useEffect(()=>{
        if(voteMenu && Array.isArray(voteMenu)) {
            document.querySelector('body').style.overflow = "hidden";
        } else {
            document.querySelector('body').style.overflow = "";
        }
    }, [voteMenu]);

    return (
        <div className={"Voteform "  + ((voteMenu && Array.isArray(voteMenu)) ? "Voteform_show" : "")}>
            <button className="Voteform__close" onClick={()=>setVoteMenu(false)}></button>
            <h2 className="subtitle">{status == "voted" ? "Результаты голосования" : "Проголосуйте за лучшую достопримечательность"}</h2>
            {
                status == "voted" ? 
                    <div className="Voteform__form">
                        {Array.isArray(voteMenu) && voteMenu.map((each, i)=>{
                            return (
                                <div className={"Voteform__form__block Voteform__form__block_e " + (each.selected !== undefined ? "Voteform__form__block_selected" : "") } key={each.city+i}>
                                    {each.city} <span>{each.count}</span>
                                </div>
                            )

                        })}
                    </div> :
                    <form className="Voteform__form" onSubmit={(e)=>{
                        e.preventDefault();
                        if(status === true && !select) return;
                        setStatus(true);
                        const formData = new FormData(e.target);
                        fetch('http://localhost/api/vote/set-vote', {method: 'POST', body: formData, cache: "no-cache"})
                            .then(data=>{
                                if(!data.ok) {
                                    throw new Error();
                                }
                                return data.json();
                            })
                            .then(data=>{
                                setStatus('voted');
                                setVoteMenu(data);
                                setNotification((current)=>({...current, message: "Ваш голос учтен"}));
                            })
                            .catch(()=>{
                                setNotification((current)=>({...current, status: false, message: 'Непредвиденная ошибка, попробуйте перезагрузить страницу.'}));
                                setVoteMenu(false);
                            })
                            
                    }}>
                        {
                            Array.isArray(voteMenu) && voteMenu.map((each, i)=>{

                                return (
                                    <label className={"Voteform__form__block " + (select == each ? "Voteform__form__block_selected" : "")} key={each+i} htmlFor={each+i}>
                                        <input className="Voteform__form__block__input" id={each+i} type="radio" name="voice" value={each}
                                            onClick={()=>setSelect(each)}/>
                                        {each}
                                    </label>
                                )
                            }) 
                        }
                        <button className="Voteform__form__submit" type="submit">{status ? <Loader/> : 'Проголосовать'}</button>
                    </form>
            }
        </div>

    )
}