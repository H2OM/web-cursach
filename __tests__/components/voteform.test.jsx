import { render, screen } from "@testing-library/react";
import customRender from "@/__mocks__/customRender";
import userEvent from "@testing-library/user-event";
import settingFetchResult from "@/__mocks__/fetchMock";
import Voteform from "@/components/voteform/client/Voteform";
import ClientContext from "@/lib/context/ClientContext";
const context = {
    voteMenu: ['testcity1','testcity2', 'testcity3'],
    setVoteMenu (val) { this.voteMenu = val},
    setNotification: (val)=>{},
}

describe('test voteform', ()=>{
    
    it('test render', async ()=>{
        const {container} = customRender(<Voteform/>, {providerProps: context});
        const btns = screen.getAllByRole('button');

        expect(btns.length).toBe(2);
        expect(screen.getByText('Проголосуйте за лучшую достопримечательность')).toBeInTheDocument();
        context.voteMenu.forEach(each=>{
            expect(screen.getByText(each)).toBeInTheDocument();
        });

        expect(container).toMatchSnapshot();
    });
});
