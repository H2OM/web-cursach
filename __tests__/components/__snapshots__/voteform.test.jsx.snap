// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test voteform test render 1`] = `
<div>
  <div
    class="Voteform Voteform_show"
  >
    <button
      class="Voteform__close"
    />
    <h2
      class="subtitle"
    >
      Проголосуйте за лучшую достопримечательность
    </h2>
    <form
      class="Voteform__form"
    >
      <label
        class="Voteform__form__block "
        for="testcity10"
      >
        <input
          class="Voteform__form__block__input"
          id="testcity10"
          name="voice"
          type="radio"
          value="testcity1"
        />
        testcity1
      </label>
      <label
        class="Voteform__form__block "
        for="testcity21"
      >
        <input
          class="Voteform__form__block__input"
          id="testcity21"
          name="voice"
          type="radio"
          value="testcity2"
        />
        testcity2
      </label>
      <label
        class="Voteform__form__block "
        for="testcity32"
      >
        <input
          class="Voteform__form__block__input"
          id="testcity32"
          name="voice"
          type="radio"
          value="testcity3"
        />
        testcity3
      </label>
      <button
        class="Voteform__form__submit"
        type="submit"
      >
        Проголосовать
      </button>
    </form>
  </div>
</div>
`;
